File: email-service/src/main/java/com/github/manosbatsis/services/email/mapper/UserMapper.java
package com.github.manosbatsis.services.email.mapper;

import com.github.manosbatsis.services.email.model.UserEvent;
import com.github.manosbatsis.services.email.model.UserEventKey;
import com.github.manosbatsis.services.user.messages.UserEventMessage;

import org.mapstruct.Mapper;
import org.springframework.context.annotation.Configuration;
import org.springframework.messaging.Message;

import java.util.Date;

@Configuration
@Mapper(componentModel = "spring")
public interface UserMapper {

    default UserEvent createUserEvent(Message<UserEventMessage> message) {
        UserEventMessage userEventMessage = message.getPayload();
        UserEvent userEvent = new UserEvent();
        userEvent.setKey(
                new UserEventKey(
                        userEventMessage.getUserId(),
                        new Date(userEventMessage.getEventTimestamp())));
        userEvent.setType(userEventMessage.getEventType().toString());
        CharSequence userJson = userEventMessage.getUserJson();
        if (userJson != null) {
            userEvent.setData(userJson.toString());
        }
        return userEvent;
    }
}

File: email-service/src/main/kotlin/com/github/manosbatsis/services/email/mapper/UserMapper.kt
package com.github.manosbatsis.services.email.mapper

import com.github.manosbatsis.lib.core.log.loggerFor
import com.github.manosbatsis.services.email.model.UserEvent
import com.github.manosbatsis.services.email.model.UserEventKey
import com.github.manosbatsis.services.user.messages.EventType
import com.github.manosbatsis.services.user.messages.UserEventMessage
import java.util.Date

object UserMapper {

    private val log = loggerFor<UserMapper>()

    fun toUserEvent(userEventMessage: UserEventMessage): UserEvent {
        log.warn("toUserEvent, userEventMessage: {}", userEventMessage)
        return UserEvent(
            key = UserEventKey(
                userEventMessage.getUserId(),
                Date(userEventMessage.getEventTimestamp()),
            ),
            type = userEventMessage.getEventType().toString(),
            data = userEventMessage.getUserJson(),
        )
    }

    fun toUserEventMessage(userEvent: UserEvent): UserEventMessage {
        log.warn("toUserEventMessage, userEvent: {}", userEvent)
        return UserEventMessage(
            userEvent.key.asString(),
            userEvent.key.datetime.time,
            EventType.valueOf(userEvent.type),
            userEvent.key.userId,
            userEvent.data,
        )
    }
}
